File: Android/app/src/main/java/com/folio/app/ui/auth/AuthViewModel.kt
package com.folio.app.ui.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.folio.app.data.api.ApiService
import com.folio.app.data.models.User
import com.folio.app.util.TokenManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class AuthViewModel @Inject constructor(
    private val api: ApiService,
    private val tokenManager: TokenManager
) : ViewModel() {

    sealed class AuthState {
        object Loading : AuthState()
        data class Authenticated(val user: User) : AuthState()
        object Unauthenticated : AuthState()
    }

    private val _authState = MutableStateFlow<AuthState>(AuthState.Loading)
    val authState: StateFlow<AuthState> = _authState

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading

    private val _verificationSent = MutableStateFlow(false)
    val verificationSent: StateFlow<Boolean> = _verificationSent

    private val _resetSent = MutableStateFlow(false)
    val resetSent: StateFlow<Boolean> = _resetSent

    init {
        checkAuth()
    }

    fun checkAuth() {
        viewModelScope.launch {
            val token = tokenManager.getAccessToken()
            if (token.isNullOrEmpty()) {
                _authState.value = AuthState.Unauthenticated
                return@launch
            }
            try {
                val resp = api.getMe()
                if (resp.isSuccessful && resp.body() != null) {
                    _authState.value = AuthState.Authenticated(resp.body()!!)
                } else {
                    tokenManager.clearTokens()
                    _authState.value = AuthState.Unauthenticated
                }
            } catch (e: Exception) {
                tokenManager.clearTokens()
                _authState.value = AuthState.Unauthenticated
            }
        }
    }

    fun login(email: String, password: String) {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null
            try {
                val resp = api.login(mapOf("email" to email, "password" to password))
                if (resp.isSuccessful) {
                    val body = resp.body()
                    body?.accessToken?.let { token ->
                        tokenManager.saveTokens(token, body.refreshToken)
                        body.user?.let { _authState.value = AuthState.Authenticated(it) }
                            ?: checkAuth()
                    }
                } else {
                    _error.value = "Неверный email или пароль"
                }
            } catch (e: Exception) {
                _error.value = e.localizedMessage ?: "Ошибка сети"
            }
            _isLoading.value = false
        }
    }

    fun register(username: String, email: String, password: String) {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null
            _verificationSent.value = false
            try {
                val resp = api.register(
                    mapOf("username" to username, "email" to email, "password" to password)
                )
                if (resp.isSuccessful) {
                    _verificationSent.value = true
                } else {
                    _error.value = "Ошибка регистрации"
                }
            } catch (e: Exception) {
                _error.value = e.localizedMessage ?: "Ошибка сети"
            }
            _isLoading.value = false
        }
    }

    fun forgotPassword(email: String) {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null
            _resetSent.value = false
            try {
                val resp = api.forgotPassword(mapOf("email" to email))
                if (resp.isSuccessful) {
                    _resetSent.value = true
                } else {
                    _error.value = "Ошибка отправки"
                }
            } catch (e: Exception) {
                _error.value = e.localizedMessage ?: "Ошибка сети"
            }
            _isLoading.value = false
        }
    }

    fun resendVerification(email: String) {
        viewModelScope.launch {
            try {
                api.resendVerification(mapOf("email" to email))
            } catch (_: Exception) {}
        }
    }

    fun logout() {
        viewModelScope.launch {
            tokenManager.clearTokens()
            _authState.value = AuthState.Unauthenticated
        }
    }

    fun clearError() { _error.value = null }
}

File: iOS/FolioApp/ViewModels/AuthViewModel.swift
import Foundation
import Combine

enum AuthState {
    case loading
    case authenticated(User)
    case unauthenticated
}

@MainActor
class AuthViewModel: ObservableObject {
    @Published var state: AuthState = .loading
    @Published var error: String?
    @Published var isLoading = false
    @Published var verificationSent = false
    @Published var resetSent = false

    var currentUser: User? {
        if case .authenticated(let user) = state { return user }
        return nil
    }

    var isAuthenticated: Bool {
        if case .authenticated = state { return true }
        return false
    }

    private let api = APIService.shared

    init() {
        Task { await checkAuth() }
    }

    func checkAuth() async {
        guard KeychainService.accessToken != nil else {
            state = .unauthenticated
            return
        }
        do {
            let user: User = try await api.get("/auth/me")
            state = .authenticated(user)
        } catch {
            KeychainService.accessToken = nil
            KeychainService.refreshToken = nil
            state = .unauthenticated
        }
    }

    func login(email: String, password: String) async {
        isLoading = true
        error = nil
        do {
            let body: [String: Any] = ["email": email, "password": password]
            let response: AuthResponse = try await api.post("/auth/login", body: body, auth: false)
            if let token = response.accessToken {
                KeychainService.accessToken = token
                if let refresh = response.refreshToken {
                    KeychainService.refreshToken = refresh
                }
                if let user = response.user {
                    state = .authenticated(user)
                } else {
                    await checkAuth()
                }
            }
        } catch let err as APIError {
            error = err.errorDescription
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    func register(username: String, email: String, password: String) async {
        isLoading = true
        error = nil
        verificationSent = false
        do {
            let body: [String: Any] = ["username": username, "email": email, "password": password]
            let _: AuthResponse = try await api.post("/auth/register", body: body, auth: false)
            verificationSent = true
        } catch let err as APIError {
            error = err.errorDescription
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    func forgotPassword(email: String) async {
        isLoading = true
        error = nil
        resetSent = false
        do {
            let body = ["email": email]
            let _: [String: String] = try await api.post("/auth/forgot-password", body: body, auth: false)
            resetSent = true
        } catch let err as APIError {
            error = err.errorDescription
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    func resendVerification(email: String) async {
        do {
            let body = ["email": email]
            let _: [String: String] = try await api.post("/auth/resend-verification", body: body, auth: false)
        } catch {}
    }

    func logout() {
        KeychainService.accessToken = nil
        KeychainService.refreshToken = nil
        state = .unauthenticated
    }
}
